### PR TITLE
Add world builder integration to CPO initial grains.

### PR DESCRIPTION
**Pull request by MFraters**
This pull request enables the use of the world builder to setup initial grain-sizes and initial grain orientations for different regions in the model.

I do still need to add tests. @Wang-yijun Can you also take a look at this to see if it works for you?

### CODE REVIEW

**Comment by Wang-yijun** [comment]
I've tested using initial textures from GWB on this branch. It works for me!

**Comment by MFraters** [comment]
@Wang-yijun Thanks for testing!

This pull request is now ready for review

File: source/particle/property/crystal_preferred_orientation.cc [hunk 1]
```diff
@@ -159,12 +162,33 @@ namespace aspect
...
+                                                                                  -this->get_geometry_model().height_above_reference_surface(position),
+                                                                                  mineral_i,
+                                                                                  n_grains);
+                double sum_volume_fractions = 0;
+                for (unsigned int grain_i = 0; grain_i < n_grains ; ++grain_i)
+                  {
+                    sum_volume_fractions += wb_grains.sizes[grain_i];
+                    volume_fractions_grains[mineral_i][grain_i] = wb_grains.sizes[grain_i];
+                    // we are receiving a array<array<double,3>,3> which needs to be unrolled in the correct way
+                    // for a tensor<2,3> it just loops first over the second index and than the first index
```
Comment (gassmoeller) [inline source/particle/property/crystal_preferred_orientation.cc:178]: This comment needs some punctuation to clarify the meaning

File: source/particle/property/crystal_preferred_orientation.cc [hunk 2]
```diff
@@ -159,12 +162,33 @@ namespace aspect
...
+                    for (unsigned int component_i = 0; component_i < 3 ; ++component_i)
+                      {
+                        for (unsigned int component_j = 0; component_j < 3 ; ++component_j)
+                          {
+                            Assert(!std::isnan(wb_grains.rotation_matrices[grain_i][component_i][component_j]), ExcMessage("Error: not a number."));
+                            rotation_matrices_grains[mineral_i][grain_i][component_i][component_j] = wb_grains.rotation_matrices[grain_i][component_i][component_j];
```
Comment (gassmoeller) [inline source/particle/property/crystal_preferred_orientation.cc:184]: Can you add a comment here that explains why you can assign the `component_i` of `wb_grains` to all `component_j` of `rotation_matrices_grains`?

File: source/particle/property/crystal_preferred_orientation.cc [hunk 3]
```diff
@@ -159,12 +162,33 @@ namespace aspect
...
+                          }
+                      }
+
+                  }
+
+                AssertThrow(sum_volume_fractions != 0, ExcMessage("Sum of volumes is equal to zero, which is not supporsed to happen. "
```
Comment (gassmoeller) [inline source/particle/property/crystal_preferred_orientation.cc:190]:
```suggestion
                AssertThrow(sum_volume_fractions != 0, ExcMessage("Sum of volumes is equal to zero, which is not supposed to happen. "
```

File: source/particle/property/crystal_preferred_orientation.cc [hunk 4]
```diff
@@ -1156,9 +1180,21 @@ namespace aspect
               prm.enter_subsection("Initial grains");
               {
                 const std::string model_name = prm.get("Model name");
-                AssertThrow(model_name == "Uniform grains and random uniform rotations",
-                            ExcMessage("No model named " + model_name + "for CPO particle property initialization. "
-                                       + "Only the model \"Uniform grains and random uniform rotations\" is available."));
+                if (model_name  == "Uniform grains and random uniform rotations")
```
Comment (gassmoeller) [inline source/particle/property/crystal_preferred_orientation.cc:1183]:
```suggestion
                if (model_name == "Uniform grains and random uniform rotations")
```

File: tests/world_builder_select_grains.prm [hunk 5]
```diff
@@ -0,0 +1,114 @@
+# Enable if: ASPECT_WITH_WORLD_BUILDER
+# This test is a copy of world_builder_simple.prm and checks that we can select
+# a subset of compositions for which to evaluate the world builder.
```
Comment (gassmoeller) [inline tests/world_builder_select_grains.prm:3]: I think this comment needs an update.

Comment (gassmoeller) [reply tests/world_builder_select_grains.prm:3]: Also, if this is just an update, can you `include` the other test .prm file?

File: tests/world_builder_select_grains/particles/particles-00000.0000.vtu
Comment (gassmoeller) [inline tests/world_builder_select_grains/particles/particles-00000.0000.vtu:1]: could you change the output format to `gnuplot` to simplify the test comparison?

**Review by gassmoeller — COMMENTED** [review]
Very nice, a few comments. 
Does this PR work cleanly with the world builder version included in ASPECT, or do you need a development version? If you need a development version we need a check for WORLD_BUILDER_VERSION in this plugin.

File: source/particle/property/crystal_preferred_orientation.cc (hunk 2 above)
Comment (MFraters) [reply source/particle/property/crystal_preferred_orientation.cc:184]: I don't understand what you mean here. component i,j is assigned to component i,j

**Comment by MFraters** [comment]
@gassmoeller I addressed your comments, but I didn't understand what you meant with one of them. This pull request works with the world builder which is part of aspect. With the changes @Wang-yijun made, better tests can be made, but that will have to wait until the new wold builder version is released and becomes part of aspect.

**Review by gassmoeller — COMMENTED** [review]
Very nice, thanks. I found a few more minor points.
Could you also add a changelog entry? As far as I see this is a new feature that was added.

**Comment by MFraters** [comment]
@gassmoeller Thanks for the review. I have addressed your comments

File: include/aspect/particle/property/crystal_preferred_orientation.h [hunk 6]
```diff
@@ -86,6 +86,17 @@ namespace aspect
         spin_tensor, drex_2004
       };
 
+      /**
+       * @brief An enum used to deterime how the initial grain sizes and orientation are set for all particles
```
Comment (gassmoeller) [inline include/aspect/particle/property/crystal_preferred_orientation.h:90]:
```suggestion
       * @brief An enum used to determine how the initial grain sizes and orientations are set for all particles
```

**Review by gassmoeller — APPROVED** [review]
One typo left to fix, afterwards feel free to merge yourself.

**Pull request merged**